src/language-markdown/printer-markdown.js

```diff
@@ -540,21 +549,173 @@ function getAncestorNode(path, typeOrTypes) {
+ * @returns {boolean} `true` if given node can be converted to space, `false` if not (i.e. newline or empty character)
+ */
+function canBeConvertedToSpace(path, value, adjacentNodes) {
+  return (
+    // "\n" or " ", of course " " always can be converted to Space
+    value !== "\n" ||
+    // no adjacent nodes
+    typeof adjacentNodes !== "object" ||
+    // "\n" between non-CJ (not han or kana) characters always can converted to Space
+    (adjacentNodes.previous?.kind !== KIND_CJ_LETTER &&
+      adjacentNodes.next?.kind !== KIND_CJ_LETTER) ||
+    (!(
+      // Do not convert it to Space when:
+      // "\n" between CJ always SHALL NOT be converted to space
+      (
+        (adjacentNodes.previous?.kind === KIND_CJ_LETTER &&
+          adjacentNodes.next?.kind === KIND_CJ_LETTER) ||
+        // Shall not be converted to Space around CJK punctuation
+        adjacentNodes.previous?.kind === KIND_CJK_PUNCTUATION ||
+        adjacentNodes.next?.kind === KIND_CJK_PUNCTUATION
+      )
+    ) &&
+      // The following rules do not precede the above rules.
+      //
+      // 1. "\n" between special signs and CJ characters
+      // [corresponding sign][Space][any string][CJ][[\n]][target sign]
+      // we wonder if there are other marks to be considered.
+      ((adjacentNodes.next.value === ":::" &&
+        isCorrespondingMarkFollowedBySpaceBefore(
+          path,
+          adjacentNodes.next.value
+        )) ||
+        // 2. If sentence uses space between CJ and alphanumerics (including hangul because of backward-compatibility),
+        //    "\n" can be converted to Space.
+        // Note: Koran uses space to divide words, so it is difficult to determine if "\n" should be converted to Space.
+        isSentenceUseCJDividingSpace(path)))
+  );
+}
+
+/**
+ * @param {*} path
+ * @param {WhiteSpaceValue} value
+ * @param {*} options
+ * @param {AdjacentNodes | undefined} [adjacentNodes]
+ */
+function printLine(path, value, options, adjacentNodes) {
   if (options.proseWrap === "preserve" && value === "\n") {
     return hardline;
   }
 
   const isBreakable =
     options.proseWrap === "always" &&
     !getAncestorNode(path, SINGLE_LINE_NODE_TYPES);
-  return value !== ""
-    ? isBreakable
-      ? line
-      : " "
-    : isBreakable
-    ? softline
-    : "";
+
+  // Space or empty
+  if (value !== "\n") {
+    return convertToLineIfBreakable(value);
+  }
+  // Chinese and Japanese does not use U+0020 Space to divide words, so U+00A0 No-break space must not be replaced with it.
+  // Behavior in other languages will not be changed because there are too much things to consider. (PR welcome)
+  return convertToLineIfBreakable(
+    canBeConvertedToSpace(path, value, adjacentNodes) ? " " : ""
+  );
+
+  /**
+   * @param value {" " | ""}
+   */
+  function convertToLineIfBreakable(value) {
+    if (!isBreakable) {
+      return value;
+    }
+    return value === " " ? line : softline;
+  }
```

Comment:
@sosukesuzuki 

Could you tell me which is better this or the below?

```js
if (value !== "" && canLineBreakBeConvertedToSpace(path, value, adjacentNodes)) {
  return isBreakable ? line : " ";
}
// Space is not allowed, just "" or "\n"
return isBreakable ? softline : "";
```

(by https://github.com/prettier/prettier/pull/11597/files/5a84ae7efec6a087cf423719f9eab127c61cb061#r962280229)

This makes it possible to remove the last line of the below thanks to `if (value !== "\n") {`.

```js
function canBeConvertedToSpace(path, value, adjacentNodes) {
  return (
    // "\n" or " ", of course " " always can be converted to Space
    value !== "\n" ||
```